### ISSUE TITLE
Make it possible to authenticate using 3rd party services

### ISSUE THREAD

**Issue by humphd** [mozilla/thimble.mozilla.org#1752]
At one point, Mozilla had a lot invested in a particular authentication model (e.g., Persona).  Today, that's not the case, and I think it's time to revisit the idea of allowing users to authenticate with Thimble using their Github, Twitter, Facebook, Google, etc. account.

We're also seeing increased interest from other people wanting to do their own installations of Thimble, and not forcing people to use our authentication scheme seems like a logical decoupling.  What would it take to start allowing something like http://passportjs.org/ in addition to our current scheme?

cc @cadecairos

**Comment by gideonthomas** [comment]
cc @simonwex on the timeline for MoFo adapting the new auth system so that it can be one of (default) the providers we use.

**Comment by Pomax** [comment]
Don't count on it being available until at least several months from now

**Comment by gideonthomas** [comment]
ouch, ok, I guess let's start by estimating work to be done in creating a passport provider for id.wm.org

**Comment by Pomax** [comment]
not so much "ouch" as "it's a hell of a lot of work and relying on it means you won't be able to get any work done in the mean time" so the easiest thing is to not rely on it. Ideally Auth0 is just one of many login systems available, since all Thimble "as a thing" should care about is that *someone* is willing to vouch for a user session. 

And then for *our* instance of Thimble we can prioritise Auth0 once it's clear how to even tap into it.

**Comment by cadecairos** [comment]
> Don't count on it being available until at least several months from now

I heard today from someone in the know that it will be "After Q3"

**Comment by gideonthomas** [comment]
@Pomax what do you suggest our default login scheme should be though?

**Comment by Pomax** [comment]
right now? webmaker login, because that's what we're still using.

**Comment by gideonthomas** [comment]
We would probably need to write a passport provider for that. Don't know how much work that is going to be (@cadecairos, are you able to estimate?)

**Comment by humphd** [comment]
How hard can it be to wrap our current stuff for passport.js?  Literally *everything* already has a provider for it.  Can we start there, and not replace what we have, but abstract it enough that we can also include others?  I'm not suggesting we throw away our current system, btw.  Just that we make it more accessible to users by offering choices.

**Comment by gideonthomas** [comment]
@humphd agree. But from what I understand, to continue to allow using id.wm.org along with wrapping what we have with passportjs, we would need to write a provider for id.wm.org so that we can actually still use it in the context of passportjs.

One way we can probably circumvent this is by using different code paths for passportjs based authentication and id.wm.org authentication (we wouldn't need to write a provider in this case). So, it would be something like:
1. Show UI to choose between id.wm.org and other passportjs providers
2. In the code, we do something like:
```
if clicked on id.wm.org
   go through our current code path for auth
otherwise
   run passportjs and use selected provider 
``` 
This way, we will basically be using two abstract authentication pathways. Later on, it should be easy enough to remove id.wm.org and use the mozilla auth system (whatever that ends up being) as a part of passportjs itself (since it will definitely have a passport provider).

One thing we will definitely need to address is how `publish.wm.org` manages users w.r.t. projects. I think it currently uses id.wm.org's usernames that we store in the publish db, so, we might need to figure out how to change that. It also uses the usernames for creating the published project links so we might need to figure out how to handle that as well (in terms of conflicts, etc.)

**Comment by humphd** [comment]
@gideonthomas I get the sense from what you're writing above that you think a passportjs provider for id would be hard.  Why is that?  Is there some complexity that I'm not aware of?

**Comment by humphd** [comment]
When I read http://mherman.org/blog/2015/09/26/social-authentication-in-node-dot-js-with-passport/#.WKm_p47avUI and similar docs, and read code like https://github.com/jaredhanson/passport-oauth2 it seems to me that writing a provider for ID should be trivial--this is literally what passport.js was designed to do.  I can't see that we'd have to change anything with regard to ID, we just abstract the way Thimble interacts with it, such that interacting with other OAuth2 providers is also possible.

**Comment by Pomax** [comment]
On a sidenote, different codepaths are hardly unheard of - even Stackoverflow gives you multiple things to click through when you sign in.

**Comment by Pomax** [comment]
@cadecairos given https://github.com/jaredhanson/passport-oauth2 does it look like we could just make that wrap id.wmo trivially? Are we using anything beyond basic oauth2?

**Comment by gideonthomas** [comment]
@cadecairos would you be able to estimate the work involved in writing a passport provider for id.wm.org (I don't think I know enough about the id server to make that estimation)?

The secondary issue about the publish server handling projects<->users still needs to be figured out though.

**Comment by cadecairos** [comment]
> @cadecairos given https://github.com/jaredhanson/passport-oauth2 does it look like we could just make that wrap id.wmo trivially? Are we using anything beyond basic oauth2?

this should be trivial to get working.

**Comment by humphd** [comment]
We like "trivial."  Maybe this is a good student bug, with some notes from you guys on what you want?

**Comment by Pomax** [comment]
@humphd @cadecairos sounds good to me - as a passport-idwmo this could be a library on the id.webmaker.org repo. How hard is it to work on id.wmo inside the vagrant collection of apps?

**Comment by humphd** [comment]
Wouldn't we do this as a new repo vs. in id?

**Comment by gideonthomas** [comment]
I doubt it is easy to work in the id.wm.org folder inside thimble because it's a git subtree, not a submodule (so it doesn't really have git history information I believe).

**Comment by Pomax** [comment]
I meant more doing this in the actual id.webmaker.org repository, in a `lib` dir of some sort, and then once we know it works, spin it off as a node module.

**Comment by humphd** [comment]
Gotcha, make sense.  I'd say we start with "whatever is easy" and refactor from there.  Before the world ends, I want to see this app allow people to login with any kind of account :)  I've been on a mission for this to happen since Webmaker 0.1.  Almost there!

**Comment by Pomax** [comment]
filed https://github.com/mozilla/id.webmaker.org/issues/431

**Comment by humphd** [comment]
@ryanwarsaw bless you for taking this bug which I've wanted fixed for 5+ years.

**Comment by ryanwarsaw** [comment]
I've noticed that I'm falling behind in tickets, I'm temporarily un-assigning myself from this one as well just in-case someone else wants to take it. I over-estimated the amount of free time I thought I was going to have in regards to recent events, I apologize for this. Hopefully I'll be back to regularly schedule programming by next month 🤞

**Comment by humphd** [comment]
Sounds good.  @flukeout one thing to ponder, based on your comments in another email: should we ponder dropping auth completely?

**Comment by ryanwarsaw** [comment]
That would probably put us in the wrong direction strategically, if our goal is growth and getting Thimble into the hands of more people. Along with that we're going to deal with exploitation and abuse, and authentication gives us other tool in the tool box to help combat this stuff.

**Comment by humphd** [comment]
RE: not having auth, consider what Glitch is doing: https://glitch.com/faq where they say:

> While we want to keep Glitch as open as possible, there are a few restrictions in place. These are there to help maintain service levels for all, and to stop someone’s accidental infinite loop somehow costing us a small fortune. The restrictions include:
>
>   Projects created by anonymous users expire after 5 days (login via GitHub to keep your projects around).

So you could allow people to publish without auth, but maybe it goes in a bucket we wipe every few days?

cc @flukeout

**Comment by ryanwarsaw** [comment]
If we remove auth completely, then how can people make sure their projects last longer than a couple of days?

**Comment by ryanwarsaw** [comment]
See: https://github.com/ryanwarsaw/passport-webmaker.

Works flawlessly, just needs to be deployed to NPM (soon) and I also need to toss all of my documentation work in there too. Probably in the next day or so. Then we can proceed with this.

**Comment by humphd** [comment]
Let's move https://github.com/ryanwarsaw/passport-webmaker to the mozilla org so it is easier to maintain.  Thank you for driving this, @ryanwarsaw.  Awesome to see this happening.

**Comment by ryanwarsaw** [comment]
@humphd I have access to create repositories on this organization but wasn't sure if there was some kind of policy for doing it, so I worked on it under my own repository instead.

**Comment by cadecairos** [comment]
@ryanwarsaw if you add me to your repo with admin permissions, I can transfer it into mozilla and set the right team permissions.

**Comment by ryanwarsaw** [comment]
Followed up with @cadecairos on this, it's been transferred to the Mozilla organization and the appropriate permissions have been set. See: https://github.com/mozilla/passport-webmaker.

**Comment by ryanwarsaw** [comment]
The finished `passport-webmaker` module has been published, an additional PR has been submitted to thimble.mozilla.org (See: #2435). Per discussion with @gideonthomas this likely won't be going anywhere anytime soon, so I'm going to unassign myself.

This work will get us a good portion of the way there, someone else will need to follow up with some migration work should we decide to support multiple providers, but this would be redundant if we switch to Auth0 (hence I didn't do it in the first place).

**Comment by gideonthomas** [comment]
Assigning to myself (@gideonthomas) to see what we can do next.
